packages/client/test/integration/Permissions.test.ts

```diff
@@ -0,0 +1,181 @@
+import { Wallet } from 'ethers'
+
+import { clientOptions, createRelativeTestStreamId, fetchPrivateKeyWithGas } from '../test-utils/utils'
+import { StreamrClient } from '../../src/StreamrClient'
+import { Stream } from '../../src/Stream'
+import { StreamPermission } from '../../src/permission'
+import { randomEthereumAddress } from 'streamr-test-utils'
+import { EthereumAddress } from 'streamr-client-protocol'
+import { StreamEndpointsCached } from '../../src/StreamEndpointsCached'
+
+jest.setTimeout(40000)
+
+describe('Stream permissions', () => {
+
+    let client: StreamrClient
+    let stream: Stream
+    let otherUser: EthereumAddress
+
+    beforeAll(async () => {
+        const wallet = new Wallet(await fetchPrivateKeyWithGas())
+        otherUser = randomEthereumAddress()
+        client = new StreamrClient({
+            ...clientOptions,
+            auth: {
+                privateKey: wallet.privateKey,
+            }
+        })
+    })
+
+    beforeEach(async () => {
+        stream = await client.createStream({
+            id: createRelativeTestStreamId(module),
+        })
+    })
+
+    describe('happy path', () => {
+        it('direct permissions', async () => {
+            await stream.grantPermissions({
+                user: otherUser,
+                permissions: [StreamPermission.PUBLISH, StreamPermission.EDIT],
+            })
+            expect(await stream.hasPermission({
+                permission: StreamPermission.PUBLISH,
+                user: otherUser,
+                allowPublic: false
+            })).toBe(true)
+            expect(await stream.hasPermission({
+                permission: StreamPermission.EDIT,
+                user: otherUser,
+                allowPublic: false
+            })).toBe(true)
+            await stream.revokePermissions({
+                user: otherUser,
+                permissions: [StreamPermission.PUBLISH, StreamPermission.EDIT],
+            })
+            expect(await stream.hasPermission({
+                permission: StreamPermission.PUBLISH,
+                user: otherUser,
+                allowPublic: false
+            })).toBe(false)
+            expect(await stream.hasPermission({
+                permission: StreamPermission.EDIT,
+                user: otherUser,
+                allowPublic: false
+            })).toBe(false)
+        })
+
+        it('inherited permissions', async () => {
+            await stream.grantPermissions({
+                public: true,
+                permissions: [StreamPermission.PUBLISH, StreamPermission.SUBSCRIBE],
+            })
+            expect(await stream.hasPermission({
+                permission: StreamPermission.PUBLISH,
+                user: otherUser,
+                allowPublic: true
```

Comment:
Could test case where `allowPublic` set to `false`. (The method should return `false`)

Reply:
Added